### PR TITLE
document how to pretty-print objects

### LINKED ISSUE #9458: There is no documentation for custom printing of user-defined types
I have a user type that carries around a BitArray. For testing purposes, when an instance of my type gets displayed in the REPL, I'd like it to look like

    MyType(01101)

instead of how it currently looks, which is

    MyType(Bool[false,true,true,false,true])

However, as far as I can tell, there is nothing in the documentation on types (http://julia.readthedocs.org/en/latest/manual/types/) that indicates how to do this. This issue is a request for such information to be included in the documentation.

From an (unofficial?) tutorial at http://en.wikibooks.org/wiki/Introducing_Julia/Types , I found I could do this

```
function Base.show(io::IO, m::MyType)
    print(io, "MyType($(join([x?'1':'0' for x in m.data])))")
end
```

(Although I'm not sure if I'm doing this the "right way", since I had to explicitly write the name of my type into the output string.) This works if I just create one instance of MyType - but if I have an array of them then it still looks like this:

```
5-element Array{MyType,1}:
 MyType(Bool[false,false,true,false,true]) 
 MyType(Bool[true,false,false,false,true]) 
 MyType(Bool[false,false,true,false,false])
 MyType(Bool[false,true,false,false,false])
 MyType(Bool[true,false,true,true,false])  
```

Thus, there seem to be some subtleties to the system for displaying custom types, which aren't documented anywhere as far as I can tell. It would be very nice if this could be explained in the documentation.

### PR DESCRIPTION
**Pull request by stevengj**
This fixes #9458.

This has come up a zillion times on julia-users ([here](https://groups.google.com/d/msg/julia-users/_CO-2XHKNQA/xjGFF5Grmu8J), [here](https://groups.google.com/d/msg/julia-users/eudLjjKiNP0/TU60_fqGnjYJ), [here](https://groups.google.com/d/msg/julia-users/eRQh51PR7Tg/bpvanIoMBQAJ), [here](https://groups.google.com/d/msg/julia-users/nYtZE7mj8qg/52sDls0iCAAJ), [here](https://groups.google.com/d/msg/julia-users/QwFzU_PbtF8/VFkm2ZFMNVgJ), [here](https://groups.google.com/d/msg/julia-users/TTuZ_UKNpPo/fLzXQTRboQUJ), [here](https://groups.google.com/d/msg/julia-users/9B8lgy24EtY/lIlTq-4zI3wJ), ....).  I can't believe we've gone this long without a manual section on it.

### CODE REVIEW

**Comment by tkelman** [comment]
These should be doctests

File: doc/manual/types.rst [hunk 1]
```diff
@@ -1255,6 +1255,78 @@ If you apply :func:`supertype` to other type objects (or non-type objects), a
...
+If we want it to display instead as ``3.0 * exp(4.0im)``, we would
+define the following method to print the object to a given output
+object ``io`` (representing a file, terminal, buffer, etcetera; see
+:ref:`man-networking-and-streams`)::
+
+    Base.show(io::IO, z::Polar) = print(io, z.r, " * exp(", z.Θ, "im)")
+
+More fine-grained control over display of ``Polar`` objects is possible.
+In particular, sometimes one wants both a verbose multi-line printing
+format, used for displaying an single object in the REPL and other interactive
```
Comment (tkelman) [inline doc/manual/types.rst]: a single object

**Comment by tkelman** [comment]
I think PR's that aren't absolutely trivial or urgent should remain open at least long enough for contributors in all time zones to get a chance to review them before they get merged.

File: doc/manual/types.rst (hunk 1 above)
Comment (stevengj) [inline doc/manual/types.rst]: fixed.

File: doc/manual/types.rst [hunk 2]
```diff
@@ -1255,6 +1255,78 @@ If you apply :func:`supertype` to other type objects (or non-type objects), a
...
+(see :ref:`man-multimedia-io`).
+
+Moreover, you can also define ``show`` methods for other MIME types in order
+to enable richer display (HTML, images, etcetera) of objects in environments
+that support this (e.g. IJulia).   For example, we can define formatted
+HTML display of ``Polar`` objects, with superscripts and italics, via::
+
+    Base.show{T}(io::IO, ::MIME"text/html", z::Polar{T}) =
+        println(io, "<code>Polar{$T}<code> complex number: ",
+                z.r, " <i>e</i><sup>", z.Θ, " <i>i</i></sup>")
```
Comment (StefanKarpinski) [inline doc/manual/types.rst]: Would it make sense to use inline HTML in Markdown to show how this looks?

Comment (stevengj) [inline doc/manual/types.rst]: Sure, I can add it to #18649 via the [`.. raw:: html` technique](http://stackoverflow.com/questions/8621042/how-to-embed-html-in-restructured-text-file).

**Comment by nalimilan** [comment]
Thanks for adding this, but I think I'm even more confused than before. Indeed the manual now says we should use the two-argument `show` method to print the compact representation. Yet, the docstring for `showcompact` says (since https://github.com/JuliaLang/julia/commit/5d7dd9c9cde762385fdfedce470dfe9822a33a1a, but before that it was `Base.limit_output`) one should use `get(io, :compact, false)` for that. That's what e.g. `Nullable` printing uses currently.

So how are these two mechanisms related? Should they be used for different things?

**Comment by vtjnash** [comment]
The tl;dr difference should be:
- 2-arg show: default representation
- show compact: hide non-essential information (such as types and type parameters)
- show limited: truncate the output, to "fit" (which may hide essential information)
- mime-show: use more verbose formatting (same information, but better layout)

**Comment by nalimilan** [comment]
Thanks. So the manual should stop recommending the two-argument form of `show` for compact representation, and advise using the `compact` attribute instead?

**Comment by vtjnash** [comment]
It seems like "unformatted" may be a better term, since describing it as "compact" is ambiguous with the `compact` attribute?

**Comment by StefanKarpinski** [comment]
Despite being semi-convinced when Jameson, Jeff and I hashed this out in person after JuliaCon, I think this current design is too complicated and hard to explain. The multiline option was easier to explain than this. We should probably come up with something clearer for the next release.

**Comment by vtjnash** [comment]
The `multiline` attribute just completely didn't work. I think we just need to be better at explaining this upfront in  the documentation for custom IO formatting. The distinction should be:
- mimetype -> specifies the layout type (e.g. do you use `\n` ("text/plain") or `<br>` ("text/html") or `\n\n` ("text/md") or `<ul></ul>` ("text/html", for a list) to make a paragraph?)
- iocontext parameters -> specifies what information is unnecessary or uses an alternate form (e.g. using color vs. bold vs. proofreaders' marks to highlight content, or whether to drop type parameters, or etc.)

**Comment by nalimilan** [comment]
There's also the issue with `print` which is documented to give an "(un-decorated) text representation" and yet calls `show` instead of `showcompact` by default... (And `@enum` which defines `showcompact` in terms of `print`...)

**Pull request merged**